Remove CodeDom dependency from S.P.DataContractSerialization.

shmao opened the pull request:
Removing CodeDom dependency from S.P.DataContractSerialization. as System.CodeDom is a higher level OOB component.

Fix #15047 

/cc: @zhenlan @weshaggard

weshaggard commented on src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/CodeTypeReference.cs:
```diff
@@ -0,0 +1,363 @@
+using System.Diagnostics;
+using System.Globalization;
+
+namespace System.Runtime.Serialization
```
You should consider potentially sharing the source file and just conditioning the namespace or even the type visibility.

shmao replied:
I've changed the fix per your suggestion.

weshaggard commented on src/Common/src/System/CodeDom/CodeObject.cs:
```diff
@@ -5,7 +5,11 @@
 using System.Collections;
 using System.Collections.Specialized;
 
+#if !Feature_Serialization
```
For consistency with our guidelines can you make this define all uppercase.

weshaggard commented on src/Common/src/System/CodeDom/CodeObject.cs:
```diff
@@ -5,7 +5,11 @@
+#endif
 {
     [Serializable]
     public class CodeObject
```
Please also condition the visibility. We want to avoid having these be public in the serialization library even if they are in a different namespace.

weshaggard commented:
Not sure why the CodeDom tests are failing in CI. I assume you are looking into that shmao.

@stephentoub would you mind having a look at this PR I wanted to make sure you are good with this pattern.

stephentoub commented:
> @stephentoub would you mind having a look at this PR I wanted to make sure you are good with this pattern.

I took a quick look.  It looks like the serialization lib is using this CodeDom support for parsing the type arguments out of a type name string?  Could we just factor out that parsing into its own helper that's then used by both CodeDom and the serialization lib, rather than sharing these whole files and using these ifdefs to control namespace and visibility?

shmao commented:
@weshaggard , yes, I'm investigating the failure with System.CodeDom.Tests.

> System.CodeDom -> D:\j\workspace\windows_nt_re---37265eab\bin\AnyOS.AnyCPU.Release\System.CodeDom\net461\System.CodeDom.dll

According to the log it seems that System.CodeDom.Tests.csproj was using the assembly from \net461 folder. I'm not sure how this is related to the fix here...

Besides, I was able to repro the failure in my local repo when building the whole corefx repo. But when I built the `System.CodeDom.Tests.builds` alone, the build succeeded and the tests passed.

shmao commented:
@ericstj would you mind having a look at the [CI failure](https://ci.dot.net/job/dotnet_corefx/job/master/job/windows_nt_release_prtest/2057/consoleFull#704284563a82fefab-f698-416f-8fca-58544c94cd4e) with System.CodeDom.Tests. It seems to me that the test project was using the assembly from \net461 which contains very few APIs. Is it expected that the test project used the assembly from \net461?

weshaggard commented:
>I took a quick look. It looks like the serialization lib is using this CodeDom support for parsing the type arguments out of a type name string? Could we just factor out that parsing into its own helper that's then used by both CodeDom and the serialization lib, rather than sharing these whole files and using these ifdefs to control namespace and visibility?

That does seem like a better approach. @shmao do you think you could take a look at doing that?

As for the CodeDom test I'm not sure what is happening there. Something in your PR must have triggered that to start happening.

shmao commented:
I understood the build failure with System.CodeDom.Tests now. The test project didn't reference System.CodeDom directly. It referenced System.CodeDom indirectly since it referenced serialization library and S.P.DataContractSerialization had dependency on System.CodeDom. After I removed  `"System.CodeDom": "4.4.0-beta-24911-02"` from S.P.DataContractSerialization, the test project then could't   CodeDom types.

I fixed the above failure by adding 'System.CodeDom' entry in common/test-runtime/project.json.

Regarding to stephentoub's suggestion, I think it might be a better approach as we wouldn't have ifdefs. But other than that, I don't see much benefit. Essentially we would still be sharing code. And refactoring the code would be a bit more risky than the current fix. I'd prefer to merge the PR as is to get rid of CodeDom dependency from S.P.DataContractSerialization before the coming `NO MERGE` weekend.

cc: @weshaggard @zhenlan

weshaggard commented:
@shmao I think I would rather factor this out into one helper class to share the code. That way we don't end up needing to adding all kinds of #ifdef's around the code which makes it difficult to maintain and understand. Yes there is some risk with factoring it out but hopefully we have some decent test coverage. 

There isn't really any rush so if you don't get it in today you can finish it next week.

shmao commented:
Per offline discussion, I merged the PR and opened #15154 for tracking refactoring the code.
